File: node-runtime/debug-logger.ts
/**
 * Debug logger for Node.js that writes to App Group container
 * Provides persistent logs accessible without unified logging
 */

import fs from 'fs';
import path from 'path';
import os from 'os';

export enum LogLevel {
    DEBUG = 'DEBUG',
    INFO = 'INFO',
    WARNING = 'WARNING',
    ERROR = 'ERROR',
    CRITICAL = 'CRITICAL'
}

export class DebugLogger {
    private readonly logDirectory: string;
    private readonly logFileName: string;
    private readonly maxLogSize: number;
    private readonly maxLogFiles: number;

    /**
     * Initialize logger with App Group container path
     * @param component Component name (e.g., "node", "ipc")
     * @param appGroupIdentifier App Group identifier
     * @param maxLogSize Maximum size in bytes before rotation (default: 1MB)
     * @param maxLogFiles Maximum number of rotated log files to keep (default: 5)
     */
    constructor(
        component: string,
        appGroupIdentifier: string = 'group.com.one.filer',
        maxLogSize: number = 1_048_576,
        maxLogFiles: number = 5
    ) {
        const homeDir = os.homedir();
        const containerPath = path.join(homeDir, 'Library/Group Containers', appGroupIdentifier);
        this.logDirectory = path.join(containerPath, 'debug');
        this.logFileName = `${component}.log`;
        this.maxLogSize = maxLogSize;
        this.maxLogFiles = maxLogFiles;

        // Create debug directory if it doesn't exist
        try {
            fs.mkdirSync(this.logDirectory, {recursive: true});
        } catch (error) {
            // Try to write to stderr as fallback
            console.error(`[DebugLogger] Failed to create log directory: ${error}`);
        }
    }

    // MARK: - Logging Methods

    public log(level: LogLevel, message: string, file?: string, line?: number): void {
        const timestamp = new Date().toISOString();
        const location = file && line ? `[${file}:${line}]` : '';
        const logLine = `[${timestamp}] [${level}] ${location} ${message}\n`;

        this.write(logLine);
    }

    public info(message: string, file?: string, line?: number): void {
        this.log(LogLevel.INFO, message, file, line);
    }

    public debug(message: string, file?: string, line?: number): void {
        this.log(LogLevel.DEBUG, message, file, line);
    }

    public warning(message: string, file?: string, line?: number): void {
        this.log(LogLevel.WARNING, message, file, line);
    }

    public error(message: string, file?: string, line?: number): void {
        this.log(LogLevel.ERROR, message, file, line);
    }

    public critical(message: string, file?: string, line?: number): void {
        this.log(LogLevel.CRITICAL, message, file, line);
    }

    // MARK: - File Operations

    private write(content: string): void {
        const logFile = path.join(this.logDirectory, this.logFileName);

        try {
            // Check if log file needs rotation
            if (this.shouldRotate(logFile)) {
                this.rotateLogFile(logFile);
            }

            // Append to current log file
            fs.appendFileSync(logFile, content);
        } catch (error) {
            // Fallback to stderr if file write fails
            console.error(`[DebugLogger] Write failed: ${error}`);
            console.error(content);
        }
    }

    private shouldRotate(logFile: string): boolean {
        try {
            const stats = fs.statSync(logFile);
            return stats.size >= this.maxLogSize;
        } catch {
            return false;
        }
    }

    private rotateLogFile(logFile: string): void {
        try {
            // Remove oldest log file if we've hit the limit
            const oldestLog = path.join(
                this.logDirectory,
                `${this.logFileName}.${this.maxLogFiles}`
            );
            if (fs.existsSync(oldestLog)) {
                fs.unlinkSync(oldestLog);
            }

            // Shift existing rotated logs
            for (let i = this.maxLogFiles - 1; i >= 1; i--) {
                const oldPath = path.join(this.logDirectory, `${this.logFileName}.${i}`);
                const newPath = path.join(this.logDirectory, `${this.logFileName}.${i + 1}`);
                if (fs.existsSync(oldPath)) {
                    fs.renameSync(oldPath, newPath);
                }
            }

            // Rotate current log to .1
            const rotatedPath = path.join(this.logDirectory, `${this.logFileName}.1`);
            fs.renameSync(logFile, rotatedPath);
        } catch (error) {
            console.error(`[DebugLogger] Rotation failed: ${error}`);
        }
    }

    // MARK: - Utility Methods

    /**
     * Get path to current log file (for testing/inspection)
     */
    public getLogFilePath(): string {
        return path.join(this.logDirectory, this.logFileName);
    }

    /**
     * Get all log files (current + rotated)
     */
    public getAllLogFiles(): string[] {
        const files = [path.join(this.logDirectory, this.logFileName)];
        for (let i = 1; i <= this.maxLogFiles; i++) {
            const rotatedFile = path.join(this.logDirectory, `${this.logFileName}.${i}`);
            if (fs.existsSync(rotatedFile)) {
                files.push(rotatedFile);
            }
        }
        return files;
    }

    /**
     * Clear all log files
     */
    public clearLogs(): void {
        for (const file of this.getAllLogFiles()) {
            try {
                fs.unlinkSync(file);
            } catch {
                // Ignore errors
            }
        }
    }
}

File: Sources/OneFiler/DebugLogger.swift
import Foundation

/// Structured debug logger that writes to App Group container
/// Provides persistent logs accessible without `log stream`
public actor DebugLogger {

    private let logDirectory: URL
    private let logFileName: String
    private let maxLogSize: Int
    private let maxLogFiles: Int

    /// Initialize logger with App Group container path
    /// - Parameters:
    ///   - component: Component name (e.g., "extension", "bridge")
    ///   - appGroupIdentifier: App Group identifier
    ///   - maxLogSize: Maximum size in bytes before rotation (default: 1MB)
    ///   - maxLogFiles: Maximum number of rotated log files to keep (default: 5)
    public init(component: String, appGroupIdentifier: String = "group.com.one.filer", maxLogSize: Int = 1_048_576, maxLogFiles: Int = 5) throws {
        guard let containerURL = FileManager.default.containerURL(forSecurityApplicationGroupIdentifier: appGroupIdentifier) else {
            throw DebugLoggerError.appGroupNotFound
        }

        let debugDir = containerURL.appendingPathComponent("debug", isDirectory: true)

        // Create debug directory if it doesn't exist
        try FileManager.default.createDirectory(at: debugDir, withIntermediateDirectories: true)

        self.logDirectory = debugDir
        self.logFileName = "\(component).log"
        self.maxLogSize = maxLogSize
        self.maxLogFiles = maxLogFiles
    }

    // MARK: - Logging Methods

    public func log(_ level: LogLevel, _ message: String, file: String = #file, function: String = #function, line: Int = #line) {
        let timestamp = ISO8601DateFormatter().string(from: Date())
        let fileName = (file as NSString).lastPathComponent
        let logLine = "[\(timestamp)] [\(level.rawValue)] [\(fileName):\(line)] \(function) - \(message)\n"

        write(logLine)
    }

    public func info(_ message: String, file: String = #file, function: String = #function, line: Int = #line) {
        log(.info, message, file: file, function: function, line: line)
    }

    public func debug(_ message: String, file: String = #file, function: String = #function, line: Int = #line) {
        log(.debug, message, file: file, function: function, line: line)
    }

    public func warning(_ message: String, file: String = #file, function: String = #function, line: Int = #line) {
        log(.warning, message, file: file, function: function, line: line)
    }

    public func error(_ message: String, file: String = #file, function: String = #function, line: Int = #line) {
        log(.error, message, file: file, function: function, line: line)
    }

    public func critical(_ message: String, file: String = #file, function: String = #function, line: Int = #line) {
        log(.critical, message, file: file, function: function, line: line)
    }

    // MARK: - File Operations

    private func write(_ content: String) {
        let logFile = logDirectory.appendingPathComponent(logFileName)

        guard let data = content.data(using: .utf8) else { return }

        // Check if log file needs rotation
        if shouldRotate(logFile) {
            rotateLogFile(logFile)
        }

        // Append to current log file
        if FileManager.default.fileExists(atPath: logFile.path) {
            if let fileHandle = try? FileHandle(forWritingTo: logFile) {
                fileHandle.seekToEndOfFile()
                fileHandle.write(data)
                try? fileHandle.close()
            }
        } else {
            try? data.write(to: logFile)
        }
    }

    private func shouldRotate(_ logFile: URL) -> Bool {
        guard let attributes = try? FileManager.default.attributesOfItem(atPath: logFile.path),
              let fileSize = attributes[.size] as? Int else {
            return false
        }
        return fileSize >= maxLogSize
    }

    private func rotateLogFile(_ logFile: URL) {
        // Remove oldest log file if we've hit the limit
        let oldestLog = logDirectory.appendingPathComponent("\(logFileName).\(maxLogFiles)")
        try? FileManager.default.removeItem(at: oldestLog)

        // Shift existing rotated logs
        for i in stride(from: maxLogFiles - 1, through: 1, by: -1) {
            let oldPath = logDirectory.appendingPathComponent("\(logFileName).\(i)")
            let newPath = logDirectory.appendingPathComponent("\(logFileName).\(i + 1)")
            try? FileManager.default.moveItem(at: oldPath, to: newPath)
        }

        // Rotate current log to .1
        let rotatedPath = logDirectory.appendingPathComponent("\(logFileName).1")
        try? FileManager.default.moveItem(at: logFile, to: rotatedPath)
    }

    // MARK: - Utility Methods

    /// Get path to current log file (for testing/inspection)
    public func getLogFilePath() -> String {
        logDirectory.appendingPathComponent(logFileName).path
    }

    /// Get all log files (current + rotated)
    public func getAllLogFiles() -> [String] {
        var files = [logDirectory.appendingPathComponent(logFileName).path]
        for i in 1...maxLogFiles {
            let rotatedFile = logDirectory.appendingPathComponent("\(logFileName).\(i)")
            if FileManager.default.fileExists(atPath: rotatedFile.path) {
                files.append(rotatedFile.path)
            }
        }
        return files
    }

    /// Clear all log files
    public func clearLogs() {
        for file in getAllLogFiles() {
            try? FileManager.default.removeItem(atPath: file)
        }
    }
}

// MARK: - Supporting Types

public enum LogLevel: String {
    case debug = "DEBUG"
    case info = "INFO"
    case warning = "WARNING"
    case error = "ERROR"
    case critical = "CRITICAL"
}

public enum DebugLoggerError: Error {
    case appGroupNotFound
    case writeError
}
